exist-core/src/main/java/org/exist/xquery/value/FunctionParameterFunctionSequenceType.java

```diff
@@ -0,0 +1,171 @@
+    }
+
+    /**
+     * shorthand if return type is unspecified
+     * @param attributeName     The name of the parameter in the <strong>FunctionSignature</strong>.
+     * @param primaryType       The <strong>Type</strong> of the parameter.
+     * @param parameterTypes    The <strong>Types</strong> of parameters the function needs to accept.
+     * @param cardinality       The <strong>Cardinality</strong> of the parameter.
+     * @param description       A description of the parameter in the <strong>FunctionSignature</strong>.
+     * @see org.exist.xquery.FunctionSignature @see Type @see org.exist.xquery.Cardinality
+     */
+    public FunctionParameterFunctionSequenceType(final String attributeName, final int primaryType, final SequenceType[] parameterTypes, final Cardinality cardinality, final String description) {
+        super(attributeName, primaryType, cardinality, description);
+        this.parameters = parameterTypes;
+        this.arity = parameterTypes.length;
+    }
+
+    /**
+     * FIXME: is this constructor needed?
+     * @param attributeName
+     */
+    public FunctionParameterFunctionSequenceType(final String attributeName) {
+        super(attributeName);
+    }
+
+    @Override
+    public boolean checkType(Sequence seq) throws XPathException {
+        // all functions?
+        if (!Type.subTypeOf(seq.getItemType(), getPrimaryType())) {
+            throw new XPathException(ErrorCodes.XPTY0004,
+                    "Type error: expected type: "
+                            + Type.getTypeName(getPrimaryType())
+                            + "; got: "
+                            + Type.getTypeName(seq.getItemType()));
+        }
+
+        // check each ref
+        FunctionReference next;
+        for (final SequenceIterator i = seq.iterate(); i.hasNext(); ) {
+            next = (FunctionReference) i.nextItem();
+            if (!checkType(next)) {
+                throw new XPathException(ErrorCodes.XPTY0004,
+                        "Type error: expected type: "
+                                + Type.getTypeName(getPrimaryType())
+                                + "; got: "
+                                + Type.getTypeName(seq.getItemType()));
+            }
+        }
+        return true;
+    }
+
+    public boolean checkType(final FunctionReference ref) throws XPathException {
+        final FunctionSignature sig = ref.getSignature();
+        // check arity of referenced function call
+        arityMatches(sig);
+
+        // check return type
+        returnTypeMatches(sig);
+
+        if (arityUnspecified()) { return true; }
+
+        // check argumentTypes
+        final SequenceType[] arguments = sig.getArgumentTypes();
+        for (int i = 0; i < arguments.length; i++) {
+            final int argumentType = arguments[i].getPrimaryType();
+            final int parameterType = parameters[i].getPrimaryType();
+            // ITEM is likely unspecified return type - catch error later
```

Comment:
Can be improved by final assignment of return type in Constructor, and using either a tombstone value to indicate non-assignment, or a null Integer instead of an int.